src/smt/theory_str_noodler/nielsen_decision_procedure.cpp

```diff
@@ -0,0 +1,567 @@
+     * @brief Get length constraints of the solution
+     *
+     * @param variable_map Mapping of BasicTerm variables to the corresponding z3 variables
+     * @return expr_ref Length formula describing all solutions (currently true)
+     */
+    expr_ref NielsenDecisionProcedure::get_lengths(const std::map<BasicTerm, expr_ref>& variable_map) {
+        if(this->init_length_sensitive_vars.size() == 0) {
+            return expr_ref(m.mk_true(), m);
+        }
+
+        std::map<BasicTerm, expr_ref> actual_var_map{};
+        expr_ref formula(this->m.mk_true(), this->m);
+        for(size_t i = 0; i < this->length_paths.size(); i++) {
+            if(this->length_paths_index >= this->length_paths[i].size()) {
+                continue;
+            }
+            expr_ref part_fle = length_formula_path(this->length_paths[i][this->length_paths_index], variable_map, actual_var_map);
+            formula = this->m.mk_and(formula, part_fle);
+        }
+        formula = this->m.mk_and(formula, generate_len_connection(variable_map, actual_var_map));
+
+        STRACE("str", tout << "Nielsen lengths: " << mk_pp(formula, this->m) << std::endl;);
+        this->length_paths_index++;
+        return formula;
+    }
+
+    /**
+     * @brief Initialize computation.
+     */
+    void NielsenDecisionProcedure::init_computation() {
+
+    }
+
+    /**
+     * @brief Preprocessing.
+     */
+    void NielsenDecisionProcedure::preprocess(PreprocessType opt) {
+        this->prep_handler = FormulaPreprocess(this->formula, this->init_aut_ass, this->init_length_sensitive_vars, m_params);
+
+        this->prep_handler.separate_eqs();
+
+        // Refresh the instance
+        this->init_aut_ass = this->prep_handler.get_aut_assignment();
+        this->init_length_sensitive_vars = this->prep_handler.get_len_variables();
+        this->formula = this->prep_handler.get_modified_formula();
+        this->formula = this->formula.split_literals();
+    }
+
+    /**
+     * @brief Compute next solution. Generate all Nielsen graphs and perform the satifiability checking.
+     * 
+     * @return True -> satisfiable
+     */
+    bool NielsenDecisionProcedure::compute_next_solution() {
+        bool sat = true;
+        // if the compute_next_solution was called for the first time
+        if(this->graphs.size() == 0) {
+            std::vector<Formula> instances = divide_independent_formula(this->formula);
+            for(const Formula& fle : instances) {
+                bool is_sat;
+                // create Nielsen graph and trim it
+                NielsenGraph graph = generate_from_formula(fle, is_sat);
+                graph = graph.trim();
+                this->graphs.push_back(graph);
+                sat = sat && is_sat;
+
+                // if some Nielsen graph is unsat --> unsat
+                if(!sat) {
+                    return false;
+                }
+
+                /// TODO: generate counter system only if it contains length variables 
```

Comment:
We can check for this already, can we not?

Reply:
Yes